pkg/dockerregistry/server/tagservice.go

```diff
@@ -150,39 +152,38 @@ func (t tagService) Lookup(ctx context.Context, desc distribution.Descriptor) ([
 }
 
 func (t tagService) Tag(ctx context.Context, tag string, dgst distribution.Descriptor) error {
-	imageStream, err := t.repo.imageStreamGetter.get()
-	if err != nil {
-		context.GetLogger(ctx).Errorf("error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
-		return distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
-	}
+	dockerImageReference := fmt.Sprintf("%s/%s/%s@%s", t.repo.config.registryAddr, t.repo.namespace, t.repo.name, dgst.Digest.String())
```

Comment:
I think this change bypasses the quota check.  How would quota be checked?

Reply:
How will access control to the image work?  Does this require the api server to have full credentials to the registry?

What are the performance implications of this change?  You're changing:

1. registry POST ImageStreamMapping to server
2. server write two objects to etcd

to

1. registry POST ImageStreamImport to server
2. server GET manifest from registry
3. registry GET ImageStream to prove client has access
4. server write two objects to etcd

Reply:
> I think this change bypasses the quota check. How would quota be checked?

The `imagestreamimport` checks the quota [1]. Also now we have a broken quota check at the time of blobs uploading. 

[1]https://github.com/openshift/origin/blob/master/pkg/image/registry/imagestreamimport/rest.go#L344

Reply:
> you're doing to double latency and number of API calls

My change adds one read from the registry (which will induce a get API call) to the write call. But in the whole it won't double number of API calls: manifest uploading is just a part of the image uploading process, and it has other steps too (layers verification, authentication).

Also, in the old code, ImageStreamMapping is created twice (from manifest service and from tag service). If it didn't not exist yet, then it would be created for the 3rd time. In the new code, only 1 API call will be preformed.

The new code gives the registry authority for untagged manifests, and solves #9550. It allows us to upload manifest lists to the integrated registry. The integrated registry becomes just a regular docker registry, and to support new types of manifests changes will be required only on API side (which seems more elegant).

This PR removes from the integrated registry responsibility for filling DockerImageMetadata. It makes the integrated registry a step closer to not using internal pkg/image/apis/image.

And in the end, the best code is no code. We can eventually remove ImageStreamMapping from master API without loosing functionality.

But if you don't like it, you can close the PR, though.